### PR TITLE
Add __torch_function__ benchmarks.

### PR DESCRIPTION
**Pull request by hameerabbasi**

### CODE REVIEW

**Comment by hameerabbasi** [comment]
Note: Do NOT merge this yet. It'll need updating after #34303 is merged.

File: benchmarks/overrides_benchmark/README.md
Comment (rgommers) [inline benchmarks/overrides_benchmark/README.md]: add "note that there are more extensive docs on using py-spy in CONTRIBUTING.md"

File: benchmarks/overrides_benchmark/README.md
Comment (rgommers) [inline benchmarks/overrides_benchmark/README.md]: I'm tempted to add a lot more context here. I can have a go at writing it. 

Interesting:
- Overhead for `torch` functions consuming tensors is on the order of `2 us`. 
- `__torch_function__` should add zero overhead for `Tensor` instances, a small overhead for tensor-like, and on the order of a couple of us for tensor-likes with `__torch_function__` method
- changing the dispatching mechanism may result in small O(100 ns) changes in overhead that are quite difficult to measure well but important.
- etc.

File: benchmarks/overrides_benchmark/README.md
Comment (rgommers) [inline benchmarks/overrides_benchmark/README.md]: need to add section `How to report benchmarking results`

Comment (rgommers) [reply benchmarks/overrides_benchmark/README.md]: it includes which function in `profile.svg` to take as top-level

File: benchmarks/overrides_benchmark/README.md
Comment (rgommers) [inline benchmarks/overrides_benchmark/README.md]: let's change this to have profiles for each of the 3 relevant cases, on master and on the feature branch

**Review by rgommers — CHANGES_REQUESTED** [review]
Still a bit too concise - the goal is really to make sure we get complete and consistent benchmarking results for future PRs that touch the dispatching code.

**Comment by hameerabbasi** [comment]
The `clang-tidy` failure is related to the network.

**Comment by ezyang** [comment]
@ngoldbaum I'll let you review first!

File: benchmarks/overrides_benchmark/README.md
Comment (ngoldbaum) [inline benchmarks/overrides_benchmark/README.md]: I think quoting exact numbers like this may not be that helpful, won't the exact normalization depend on which CPU was used?

File: benchmarks/overrides_benchmark/README.md
Comment (ngoldbaum) [inline benchmarks/overrides_benchmark/README.md]: Couldn't this part be automated? Maybe rather than relying on the `py-spy` flame graph svg output it would make more sense to process the data from the "raw" output format to extract relevant information that can be copy/pasted into a PR discussion.

File: torch/csrc/utils/python_arg_parser.cpp
Comment (ngoldbaum) [inline torch/csrc/utils/python_arg_parser.cpp]: I don't think *all* changes to the argument parser should require running the benchmarks.

File: benchmarks/overrides_benchmark/common.py
Comment (ngoldbaum) [inline benchmarks/overrides_benchmark/common.py]: Maybe also a test with a tensor-like class that doesn't inherit from `torch.Tensor` and implements `__torch_function__`.

File: benchmarks/overrides_benchmark/README.md
Comment (hameerabbasi) [reply benchmarks/overrides_benchmark/README.md]: Are there docs on that format somewhere?

File: benchmarks/overrides_benchmark/README.md
Comment (hameerabbasi) [reply benchmarks/overrides_benchmark/README.md]: I'll change the wording here, but @rgommers requested we put numbers in.

Comment (rgommers) [reply benchmarks/overrides_benchmark/README.md]: I think I said "on the order of 2 us". Which is useful imho. Numbers don't vary _that_ widely, typically by 20-40%, only if you have a really old CPU do you get larger times.

The point is to somehow express that we're looking for O(100 ns) numbers that are really hard to benchmark but still significant. Doing that without mentioning any numbers at all will be more convoluted.

File: torch/csrc/utils/python_arg_parser.cpp
Comment (rgommers) [reply torch/csrc/utils/python_arg_parser.cpp]: Agreed, can change to "If changing this file in a way that may affect the __torch_function__ overhead"

File: benchmarks/overrides_benchmark/README.md
Comment (ngoldbaum) [reply benchmarks/overrides_benchmark/README.md]: It’s mentioned in the output of `py-spy record —help`

Comment (hameerabbasi) [reply benchmarks/overrides_benchmark/README.md]: I mean the actual format that `raw` produces, is there a parser or some docs on it?

Comment (ngoldbaum) [reply benchmarks/overrides_benchmark/README.md]: No I don't think it's documented, py-spy's docs are pretty barebones. Looking at the py-spy source code it looks like it's writing the contents of the `HashMap` mapping stack frame names to counts to the output file, see:

https://github.com/benfred/py-spy/blob/d870b3768353001b31ee78129a79be99319b9bec/src/main.rs#L135-L145

https://github.com/benfred/py-spy/blob/d870b3768353001b31ee78129a79be99319b9bec/src/flamegraph.rs#L39-L42

Comment (benfred) [reply benchmarks/overrides_benchmark/README.md]: The py-spy raw output is in the 'folded stack' format used by flamegraph.pl (https://github.com/brendangregg/FlameGraph#2-fold-stacks) . The format is a text file with a stack trace /sample count per line, with each frame in the stack trace being delimited with a ';'.   Fwiw for benchmarking, you can take a 'before' and 'after' version of these raw files and compute a [differential flamegraph](http://www.brendangregg.com/blog/2014-11-09/differential-flame-graphs.html).

Comment (hameerabbasi) [reply benchmarks/overrides_benchmark/README.md]: I looked into this. It would introduce another dependency to rasterize the SVG, and another to diff the flame graph. I wonder if it's worth it.

Comment (hameerabbasi) [reply benchmarks/overrides_benchmark/README.md]: Not to mention auto checkout and auto builds -- Which is bad.

File: benchmarks/overrides_benchmark/common.py
Comment (hameerabbasi) [reply benchmarks/overrides_benchmark/common.py]: Done.

**Comment by ezyang** [comment]
I'm happy to merge this when y'all are OK with it. Note that you should consider running the bench as part of our CI to make sure it doesn't bitrot. (Don't have to look at the numbers though)

**Comment by hameerabbasi** [comment]
> Note that you should consider running the bench as part of our CI to make sure it doesn't bitrot. (Don't have to look at the numbers though)

Where would I make these changes? And is it okay if it takes O(10 s) to run all of them?

**Comment by ezyang** [comment]
Reduce the number of reps.

To add new scripts to CI look at `.jenkins/pytorch/test.sh`

**Comment by hameerabbasi** [comment]
CI failures seem completely unrelated.

**Review by ezyang — APPROVED** [review]

**Comment by hameerabbasi** [comment]
Looks like internal tests are failing -- Anything I can do to help fix this?

**Comment by ezyang** [comment]
they're fake, don't worry

**Comment by ezyang** [comment]
Non-ASCII characters broke tests:

```
Mar 26 20:13:22 + python bench.py -n 1 -m 1
Mar 26 20:13:22 ~/workspace/benchmarks/overrides_benchmark ~/workspace
Mar 26 20:13:22 Traceback (most recent call last):
Mar 26 20:13:22   File "bench.py", line 67, in <module>
Mar 26 20:13:22     main()
Mar 26 20:13:22   File "bench.py", line 61, in main
Mar 26 20:13:22     t.__name__, (10 ** 6) * bench_min, (10 ** 6) * bench_std,
Mar 26 20:13:22 UnicodeEncodeError: 'ascii' codec can't encode character '\u03bc' in position 54: ordinal not in range(128)
```

https://app.circleci.com/pipelines/github/pytorch/pytorch/146972/workflows/7596773d-d427-41ed-a007-48a282e155d1/jobs/4963663

**Pull request closed without merging**
